### PR TITLE
Fixed issue [security] #18553: SQL injection on surveygroup update

### PR DESCRIPTION
**Pull request by Shnoulle**
Dev: need admin right
Dev: remove usage of App()->request in model
Dev: binding param for gsid

### CODE REVIEW

File: application/models/TemplateConfiguration.php [hunk 1]
```diff
@@ -472,12 +472,13 @@ public function search()
...
     /**
-     * @todo document me
+     * Retrieves a list of models based on the current search/filter conditions.
      *
+     * @param integer $gsid Survey group, else get global
      * @return CActiveDataProvider
      * @throws Exception
      */
-    public function searchGrid()
+    public function searchGrid($gsid = null)
```
Comment (apacheborys) [inline application/models/TemplateConfiguration.php:481]: Perhaps will be better to add here typehint instead param annotation?

File: application/models/TemplateConfiguration.php [hunk 2]
```diff
@@ -492,11 +493,11 @@ public function searchGrid()
         $criteria->addCondition('t.sid IS NULL');
         $criteria->addCondition('template.name IS NOT NULL');
 
-        // check if survey group id is present
-        $gsid = App()->request->getQuery('id', null);
         if ($gsid !== null) {
-            $criteria->addCondition('t.gsid = ' . $gsid);
+            /* Group configuration */
+            $criteria->compare('t.gsid', $gsid);
```
Comment (apacheborys) [inline application/models/TemplateConfiguration.php:498]: Please correct me if I'm wrong, here we just comparing? Do we need any additional action here?

File: application/views/admin/surveysgroups/update.php [hunk 3]
```diff
@@ -45,7 +45,14 @@
             <div id="templateSettingsFortThisGroup" class="tab-pane">
                 <?php
                     if (is_a($templateOptionsModel, 'TemplateConfiguration')){
-                        Yii::app()->getController()->renderPartial('/themeOptions/surveythemelist', array( 'oSurveyTheme'=> $templateOptionsModel, 'pageSize'=>$pageSize ));
+                        Yii::app()->getController()->renderPartial(
+                            '/themeOptions/surveythemelist',
+                            array(
+                                'oSurveyTheme' => $templateOptionsModel,
+                                'pageSize' => $pageSize,
+                                'SurveysGroup' => $model
```
Comment (apacheborys) [inline application/views/admin/surveysgroups/update.php:53]: What is the idea behind to put SurveysGroup additionally?

File: application/models/TemplateConfiguration.php (hunk 2 above)
Comment (Shnoulle) [reply application/models/TemplateConfiguration.php:498]: What do you mean ? 

compare do 

- condition = "t.gsid = :gsid"
- params [":gsid" => $gsid)

https://github.com/yiisoft/yii/blob/8530c30959d5f68c0d1b0d611e3ce2bf32a8371a/framework/db/schema/CDbCriteria.php#L449-L450

File: application/models/TemplateConfiguration.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/TemplateConfiguration.php:481]: integer or null ? 
https://www.php.net/manual/en/language.types.declarations.php#language.types.declarations.nullable

Seems OK (PHP7.1)

File: application/views/admin/surveysgroups/update.php (hunk 3 above)
Comment (Shnoulle) [reply application/views/admin/surveysgroups/update.php:53]: **This is the fix in fact**

**Comment by Shnoulle** [comment]
Seems there still an issue ? Unsure : with `/admin/surveysgroups/sa/update?id=1)%20AND%20GTID_SUBSET(CONCAT(0x717a707071%2C(SELECT%20MID((IFNULL(CAST(grantee%20AS%20NCHAR)%2C0x20))%2C1%2C190)%20FROM%20INFORMATION_SCHEMA.USER_PRIVILEGES%20LIMIT%2026%2C1)%2C0x7162707871)%2C2772)--%20cDQz`

the list of survey are empty …

**Comment by Shnoulle** [comment]
Removing getRequest in Survey::model too , need to reset group to null when search

File: application/models/TemplateConfiguration.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/TemplateConfiguration.php]: Added , BUT : here : the value are set by dev: if it's not a integer or null : think we must send a 500 error in all condition : 500 error : there are a development broken here ;)

File: application/models/Survey.php [hunk 4]
```diff
@@ -186,7 +186,11 @@ class Survey extends LSActiveRecord implements PermissionInterface
     public function init()
     {
         /** @inheritdoc */
-
+        /* Do not set any default when search, reset gsid */
+        if ($this->scenario == 'search') {
```
Comment (Shnoulle) [inline application/models/Survey.php:190]: This part improve a lot https://bugs.limesurvey.org/view.php?id=18556 !
And surely https://bugs.limesurvey.org/view.php?id=18473

File: application/models/TemplateConfiguration.php (hunk 2 above)
Comment (apacheborys) [reply application/models/TemplateConfiguration.php:498]: ok, got it, thank you

File: application/views/admin/surveysgroups/update.php (hunk 3 above)
Comment (apacheborys) [reply application/views/admin/surveysgroups/update.php:53]: ok, thank you for clarification

**Review by apacheborys — COMMENTED** [review]
LGFM

File: application/models/TemplateConfiguration.php [hunk 5]
```diff
@@ -783,9 +784,9 @@ class="btn btn-primary btn-block selector--ConfirmModal">
             href="' . $sUninstallUrl . '"
             data-post=\'{ "templatename": "' . $this->template_name . '" }\'
             data-text="' . gT('This will reset all the specific configurations of this theme.') . '<br>' . gT('Do you want to continue?') . '"
-            data-button-no="' . gt('Cancel') . '"
```
Comment (olleharstedt) [inline application/models/TemplateConfiguration.php:786]: For future, can you fix your editor to not apply whitespace changes automatically, please? It makes code review harder. :(

**Comment by olleharstedt** [comment]
@apacheborys Did you also test this fix by Denis, or only code review?

File: application/models/TemplateConfiguration.php (hunk 5 above)
Comment (Shnoulle) [reply application/models/TemplateConfiguration.php:786]: Yes, sorry …

**Comment by Shnoulle** [comment]
@olleharstedt wait a minute : i merge master to get tester

**Comment by Shnoulle** [comment]
@olleharstedt must fix `   │ Facebook\WebDriver\Exception\NoSuchElementException: Unable to locate element: #expires_datetimepicker .picker-switch a[data-action="clear"]`

since merging …

**Comment by olleharstedt** [comment]
I saw, I notified Gabriel about it.

**Comment by Shnoulle** [comment]
> I saw, I notified Gabriel about it.

It's a new feature in develop ?

**Comment by olleharstedt** [comment]
Yeah, marking users with an expiration date.

**Comment by Shnoulle** [comment]
All done …

**Pull request merged**